packages/machine/src/instruction-executor.ts

```diff
@@ -64,46 +65,64 @@ export class InstructionExecutor implements Observable {
   public buildExecutionsFromLog(log: Log): ActionExecution[] {
     return Object.keys(log).map(key => {
       const entry = log[key];
-      const action = new Action(
-        entry.requestId,
-        entry.actionName,
-        entry.clientMessage,
-        entry.isAckSide
-      );
       const execution = new ActionExecution(
-        action,
+        entry.actionName,
+        instructionGroupFromProtocolName(entry.actionName, entry.isAckSide),
         entry.instructionPointer,
         entry.clientMessage,
-        this
+        this,
+        entry.isAckSide,
+        entry.requestId,
+        entry.intermediateResults
       );
-      execution.results = entry.results;
-      action.execution = execution;
       return execution;
     });
   }
 
   public receiveClientActionMessageAck(
     msg: cf.legacy.node.ClientActionMessage
   ) {
-    this.execute(new Action(msg.requestId, msg.action, msg, true));
+    this.execute(
+      new ActionExecution(
```

Comment:
@IIIIllllIIIIllllIIIIllllIIIIllllIIIIll Following up on my previous comment, this would probably be more readable with a configuration object.